[Day3]

terry-yes opened the pull request:
- UIApplication.willResignActiveNotification과 UIApplication.didBecomeActiveNotification의 Notification을 받을 수 있게 observer를 달았습니다.
- 각 노티가 발생할 때 PurchaseListViewModel이 UserDefaults에 저장할 수 있도록 하였습니다!

![](https://images.velog.io/images/taelee/post/7b9a5207-38aa-4ba2-9e4c-e136618249c5/Jan-26-2022%2011-39-24.gif)

terry-yes committed:
feat: 웹툰 목록 화면의 상단 뷰와 하단 테이블 뷰의 간격을 20에서 0으로 수정

terry-yes committed:
chore: 커스텀셀을 CustomCells폴더로 옮기고 웹툰 목록화면의 VC이름을 MainViewController로 변경

terry-yes committed:
feat: Purchasable 프로토콜 생성

- Purchasable을 채택할 PurchaseListViewModel의 메소드명 변경(create-> makePurchase, purchase(at:)-> retrievePurchase(at:)

terry-yes committed:
chore: .DS_Store .gitignore추가

terry-yes committed:
feat: 구매기록 화면에 저장된 구매기록이 표현되도록 구현

terry-yes committed:
feat: 구매기록 화면에서 Clear버튼 누르면 모든 구매기록이 삭제되도록 구현

terry-yes committed:
fix: 메인화면 하단 테이블뷰의 오토레이아웃 변경

- safe area의 하단에 주던 오토레이아웃을 상위뷰에 주는 것으로 변경

terry-yes committed:
feat: 구매기록 화면에서 셀을 스와이프해서 구매기록을 삭제할 수 있도록 구현

terry-yes committed:
feat: 웹툰 목록 화면에서 테이블 뷰 셀이 선택되지 않도록 변경

terry-yes committed:
feat: 웹툰 목록 화면에서 이미 구매한 웹툰은 재구매가 안 되도록 구현

terry-yes committed:
feat: 앱이 백그라운드로 갈 때와 포어그라운드로 올 때 Notification Center에서 감지할 수 있도록 구현

terry-yes committed:
docs: 현재까지 구현 사항 README.md에 정리

terry-yes committed:
fix: readme 이미지 주소 변경

terry-yes committed:
feat: 앱을 끄거나 background에 갈 때 구매기록이 UserDefaults에 저장되도록 구현

terry-yes committed:
docs: 현재까지 구현사항 README.md에 기록

bamiboo-han commented on PageApp/PageApp/ViewModels/PurchaseViewModel.swift:
```diff
@@ -7,23 +7,52 @@
-    func purchase(at index: Int) -> Purchase? {
-        return index <= purchaseCount ? purchases[index] : nil
+    func retrievePurchase(at index: Int) -> PurchaseViewModel? {
+        return index <= purchaseCount ? PurchaseViewModel(purchases[index]) : nil
```
index 의 유효성 검사를 위해서 `purchases.indices.contains(index)` 를 사용하는 방법도 있으니 참고해보세요~
이 경우 index 에 0보다 작은 값도 같이 걸러지니 더 명확해집니다 :)

terry-yes replied:
오! 이런 기능을 찾고 있었는데 감사합니다
리뷰 감사드려요~😄
해당 사항 수정했습니다!

terry-yes committed:
refactor: 코드리뷰 사항 반영

- 배열내의 인덱스인지 검사하기 위해 array.indices.contains(index)사용

kyoo-sl commented on PageApp/PageApp/Base.lproj/Main.storyboard:
```diff
@@ -11,7 +11,7 @@
+                <viewController id="BYZ-38-t0r" customClass="MainViewController" customModule="PageApp" customModuleProvider="target" sceneMemberID="viewController">
                     <view key="view" contentMode="scaleToFill" id="8bC-Xf-vdC">
                         <rect key="frame" x="0.0" y="0.0" width="414" height="896"/>
                         <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
```
여기 pr이랑 상관 없긴한데, storyboard에서 상단배너의 이슈가 있는 것 같아요~ 한번 수정해보면 좋을 것 같아요~

<img width="808" alt="Screen Shot 2022-01-28 at 4 20 32 PM" src="https://user-images.githubusercontent.com/97584466/151504342-5e9125af-694c-4553-8a7d-b635ac4500f2.png">

terry-yes replied:
말씀해주신 워닝 이외에 하단에 있는 Trailing constraint 워닝과 Fixed width constraints 워닝 또한 전부 해결했습니다!

kyoo-sl reviewed (approved):
이 외 이슈 없습니다~

terry-yes committed:
refactor: 코드 리뷰 수정사항 반영

- WebtoonTableViewcell의 Expression implicitly coerced from '[String : WebtoonViewModel?]' to '[AnyHashable : Any]' 워닝 해결
- MainViewController에서 Cell에 데이터를 전달할 때 PuchaseListViewModel 전체를 전달하던 것에서 해당 셀에 대한 Puchase기록만 보내는 것으로 변경
- Storyboard에 몇가지 AutoLayout 워닝들 전부 해결
